src/transport/raw/TCP.cpp

```diff
@@ -47,13 +47,14 @@ constexpr int kListenBacklogSize = 2;
 /**
  *  Determine if the given buffer contains a complete message
  */
-bool ContainsCompleteMessage(const System::PacketBufferHandle & buffer, uint8_t ** start, uint16_t * size)
+bool ContainsCompleteMessage(const System::PacketBufferHandle & buffer, uint8_t ** start, uint16_t * size, uint16_t * payloadSize)
 {
     bool completeMessage = false;
 
     if (buffer->DataLength() >= kPacketSizeBytes)
     {
         *size           = LittleEndian::Get16(buffer->Start());
+        *payloadSize    = static_cast<uint16_t>(*size + kPacketSizeBytes);
```

Comment:
So... We should consider using a wider type here.  Otherwise this addition can overflow (keep in mind that `size` is provided in the packet, so could be literally anything) and then we'll think payloadSize is 0 or 1 while we think size is 0xFFFF/0xFFFE respectively, and it gets a bit hard to reason about the code.

That said, it's possible that all the APIs we end up passing payloadSize to take `uint16_t`, in which case I am not quite sure how to best handle things in this situation....  Maybe immediately check for it being > UINT16_MAX, and if so ... I'm not sure.  We'll never be able to stick that many bytes in a packetbuffer chain (because TotalLength() returns `uint16_t`), right?

If the answer is that if someone is malicious like this our connection to them (and only them) gets stuck until it times out, that's probably fine.  We should just make sure we don't end up reaching any code that would go awry in other ways in the overflow situation....

Followup is probably OK for this.

Reply:
Let's see. If there is `payloadSize` overflow, the message must be invalid, and `completeMessage` will be false (IP message length is 16 bits, so DataLength() will never be larger).

So we can check for overflow (`payloadSize < messageSize`) below and consume… what? We know the `messageSize` is wrong, but we don't know the actual TCP message length. (At the LwIP level it's TotalLength, but that gets lost when in mRcvQueue.)

Is there a bigger problem here, trusting that a non-overflowing `messageSize` is correct? If it's ever wrong, then the next place we look for a messageSize will be the middle of a packet somewhere, with an incorrect messageSize…

Reply:
Well yes.  If messages are only length-delimited, then any delimiter error will permanently corrupt a stream.  But I think this is just something one accepts.  We trust that the TCP stream itself is correct and uncorrupted.  If the sender sends garbage, then things will break.  That's also true for datagrams.

Re: type.  Yes, I think @bzbarsky-apple has a good point.  This is what size_t is for.  That is guaranteed to work internally.  Then all we need is a static assert for `sizeof(size_t) >= sizeof(uint16_t)` and runtime checks when writing into the message's uint16_t length field.

Reply:
Right, to be clear, if `messageSize` is bogus then the only things we can do are to either accept it at face value and consume that many bytes from the stream or drop the TCP connection and not process anything else from the stream.  I don't see any other real options here.

My point about overflow is that we should not have different parts of the code have different ideas of how much message we are trying to consume (e.g. thinking that our entire TCP communication is 1 byte but that we have 0xFFFE bytes of CHIP data in there).  That's just a recipe for us getting our internal state confused somehow in likely-bad ways.  Or at least that's my prior on this sort of situation....